### PR TITLE
fix: (helm) - in some rare helm install cases, status is missing InstallSuccessful

### PR DESCRIPTION
**Pull request by mikeshng**
Signed-off-by: Mike Ng <ming@redhat.com>

**Description of the change:**
In some rare cases, the InstallSuccessful reason is not stamped to the helm operator based CR after a successful helm install.
This fix is to make sure that it will keep trying to put the InstallSuccessful into the status.

**Motivation for the change:**
Closes https://github.com/operator-framework/operator-sdk/issues/3728

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [Not applicable] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [Not applicable] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/helm/controller/reconcile.go [hunk 1]
```diff
@@ -225,6 +225,21 @@ func (r HelmOperatorReconciler) Reconcile(request reconcile.Request) (reconcile.
...
 		return reconcile.Result{RequeueAfter: r.ReconcilePeriod}, err
+	} else if !manager.IsUpgradeRequired() &&
+		!status.ContainsCondition(types.ConditionDeployed) &&
+		!status.ContainsCondition(types.ConditionReleaseFailed) {
+		status.SetCondition(types.HelmAppCondition{
+			Type:    types.ConditionDeployed,
+			Status:  types.StatusTrue,
+			Reason:  types.ReasonInstallSuccessful,
+			Message: manager.ReleaseMessage(),
+		})
```
Comment (joelanford) [inline internal/helm/controller/reconcile.go:236]: Can you move this to line ~334? We should just _always_ set the condition.

At that point, we can also look at `expectedRelease`, so I don't think we need the extra manager functions.

Lastly, we should look at the release version to know whether to set `ReasonInstallSuccessful` or `ReasonUpgradeSuccessful`.

WDYT? Will the problem be still be solved if those tweaks are made?

**Comment by joelanford** [comment]
Also, I think this would be a bug fix that should have a changelog fragment. Would you mind adding that as well?

File: changelog/fragments/helm-install-status-fix.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+entries:
+  - description: Fixed a bug that caused the Helm operator not to set the `InstallSuccessful` status.
```
Comment (camilamacedo86) [inline changelog/fragments/helm-install-status-fix.yaml:2]:
```suggestion
  - description: Ensures that the `InstallSuccessful` status will be set always by checking if the release info. 
```

Comment (camilamacedo86) [reply changelog/fragments/helm-install-status-fix.yaml:2]: IHMO ^

Comment (joelanford) [reply changelog/fragments/helm-install-status-fix.yaml:2]: I think @mikeshng's version reads more clearly with one tweak though.
```suggestion
  - description: Fixed a bug that caused the Helm operator not to set the `InstallSuccessful` and `UpgradeSuccessful` status reasons when the status update fails during installation and upgrade.
```

Comment (camilamacedo86) [reply changelog/fragments/helm-install-status-fix.yaml:2]: i liked your latest version 👍

File: internal/helm/controller/reconcile.go [hunk 3]
```diff
@@ -316,6 +316,16 @@ func (r HelmOperatorReconciler) Reconcile(request reconcile.Request) (reconcile.
 	}
 
 	log.Info("Reconciled release")
+	reason := types.ReasonUpgradeSuccessful
+	if expectedRelease.Version == 1 {
+		reason = types.ReasonInstallSuccessful
+	}
```
Comment (camilamacedo86) [inline internal/helm/controller/reconcile.go:322]: Now it makes more sense for me 👍

File: internal/helm/controller/reconcile.go [hunk 4]
```diff
@@ -316,6 +316,16 @@ func (r HelmOperatorReconciler) Reconcile(request reconcile.Request) (reconcile.
...
+	status.SetCondition(types.HelmAppCondition{
+		Type:    types.ConditionDeployed,
+		Status:  types.StatusTrue,
+		Reason:  reason,
+		Message: expectedRelease.Info.Notes,
```
Comment (camilamacedo86) [inline internal/helm/controller/reconcile.go:327]: we need to check if the info is not nil. Such as : https://github.com/operator-framework/operator-sdk/blob/ed95c5af3e937a71e87bca6348bd8ad3a965e7ac/internal/helm/controller/reconcile.go#L212-L215

File: internal/helm/controller/reconcile.go (hunk 3 above)
Comment (mikeshng) [reply internal/helm/controller/reconcile.go:322]: When I ran the `helm ls` command it showed:
```
NAME        	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART      	APP VERSION
nginx-sample	default  	1       	2020-08-13 14:09:57.443165995 -0400 EDT	deployed	nginx-0.1.0	1.16.0  
``` 

So I thought `Version` is `APP VERSION` which means its not going to be `1` for fresh install case. Then I double checked and it seems like `Version` is infact the `REVISION` from the output above so I changed it back to what Joe suggested earlier.

File: internal/helm/controller/reconcile.go (hunk 4 above)
Comment (mikeshng) [reply internal/helm/controller/reconcile.go]: Thanks for the good catch.

File: internal/helm/internal/types/types.go [hunk 5]
```diff
@@ -129,6 +129,18 @@ func (s *HelmAppStatus) RemoveCondition(conditionType HelmAppConditionType) *Hel
...
+// Otherwise, return false.
+func (s *HelmAppStatus) ContainsCondition(conditionType HelmAppConditionType) bool {
+	for i := range s.Conditions {
+		if s.Conditions[i].Type == conditionType {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (joelanford) [inline internal/helm/internal/types/types.go:143]: This function is no longer necessary, right?

**Pull request merged**